raiden/tests/unit/fuzz/test_state_changes.py

```diff
@@ -0,0 +1,297 @@
+
+        self.chain_state = ChainState(
+            self.random,
+            self.block_number,
+            self.address,
+            factories.UNIT_CHAIN_ID,
+        )
+
+        self.token_network_id = factories.make_address()
+        self.token_id = factories.make_address()
+        self.token_network_state = TokenNetworkState(self.token_network_id, self.token_id)
+
+        self.payment_network_id = factories.make_payment_network_identifier()
+        self.payment_network_state = PaymentNetworkState(
+            self.payment_network_id,
+            [self.token_network_state],
+        )
+
+        self.chain_state.identifiers_to_paymentnetworks[
+            self.payment_network_id
+        ] = self.payment_network_state
+
+        self.channels = list()
+
+        for partner_address in self.channels_with:
+            channel = factories.make_channel(
+                our_balance=1000,
+                partner_balance=1000,
+                token_network_identifier=self.token_network_id,
+                our_address=self.address,
+                partner_address=partner_address,
+            )
+            channel_new_state_change = ContractReceiveChannelNew(
+                factories.make_transaction_hash(),
+                self.token_network_id,
+                channel,
+                self.block_number,
+            )
+            node.state_transition(self.chain_state, channel_new_state_change)
+
+            self.channels.append(channel)
+
+    def event(self, description):
+        """ Wrapper for hypothesis' event function.
+
+        hypothesis.event raises an exception when invoked outside of hypothesis
+        context, so skip it when we are replaying a failed path.
+        """
+        if not self.replay_path:
+            event(description)
+
+    @precondition(lambda self: self.channels)
+    @invariant()
+    def channel_state_invariants(self):
+        for netting_channel in self.channels:
+            our_state = netting_channel.our_state
+            partner_state = netting_channel.partner_state
+
+            our_transferred_amount = 0
+            if our_state.balance_proof:
+                our_transferred_amount = our_state.balance_proof.transferred_amount
+                assert our_transferred_amount >= 0
+
+            partner_transferred_amount = 0
+            if partner_state.balance_proof:
+                partner_transferred_amount = partner_state.balance_proof.transferred_amount
+                assert partner_transferred_amount >= 0
+
+            assert channel.get_distributable(our_state, partner_state) >= 0
+            assert channel.get_distributable(partner_state, our_state) >= 0
+
+            our_deposit = netting_channel.our_total_deposit
+            our_amount_locked = channel.get_amount_locked(our_state)
+            our_balance = channel.get_balance(our_state, partner_state)
+            assert 0 <= our_amount_locked <= our_balance <= our_deposit
```

Comment:
This is not correct:

- `our_amount_locked` may be larger then `our_deposit`, this happens when we receive a transfer partner, which gives us a balance higher then our deposit.
- `our_balance` can too be higher then our deposit, for the same reason

Reply:
Yes, and as mentioned in https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints, we have the following invariant: `(5.1 R) L1 <= AB1, L1 <= TAD, L1 >= 0`. Therefore, the above assertion can be split into:

```
assert 0 <= our_amount_locked <= our_balance
assert our_amount_locked <= our_deposit + partner_deposit  
```
Note that `TAD` is actually the total available deposit - if we had the withdraw implemented, we would instead have: `assert our_amount_locked <= our_deposit + partner_deposit - out_withdrawn - partner_withdrawn`. This MUST be changed at that point, so maybe worth documenting this in the [withdraw issue](https://github.com/raiden-network/raiden/issues/1498)?

Reply:
Thanks, I changed the checks accordingly.